feat(workflow): Add created_by to Rule and AlertRuleSerializer

iProgramStuff opened the pull request:
We'd like to display the user who created an alert in the alert rule listing. This information will be available in the [Alert]RuleActivity model data, so I've added a created_by property to the model and spit that out in the serializer.

iProgramStuff committed:
Adding some activity logging

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/trackalertrulecreator

iProgramStuff committed:
Adding user context to update

iProgramStuff committed:
Adding user context to update

iProgramStuff committed:
Adding created_by to serializer

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/trackalertrulecreator

wedamija commented on src/sentry/models/rule.py:
```diff
@@ -58,6 +58,15 @@ def get_for_project(cls, project_id):
+    def created_by(self):
+        created_activity = RuleActivity.objects.filter(
+            rule=self, type=RuleActivityType.CREATED.value
+        ).first()
```
We generally try to avoid `first` because it forces a sort order on the results that you might not expect. If you only ever expect there to be one row, then `get` is fine. If there might be more, sort  by `id` and then `.first` should be fine. We might need to change this anyway since we'll need to fetch in bulk in the serializer.

iProgramStuff replied:
I agree about fetching in bulk for the serializer - but I think I will leave this in so it could be used as a sort field.

wedamija replied:
If we sort on this then it'll have a similar problem - we need to fetch each row before we can sort. I think we'll need a way to prefetch this for a list of objects so that we only do this logic if we haven't already fetched the cached value.

iProgramStuff replied:
Aah I see. I can change it so that we can somehow prefetch these in bulk for sorting as part of another PR (like you suggested in another comment). We don't currently have a direct use-case, but I'm thinking of leaving this function in (changing it's implementation away from .first and towards .get in a try/except) for the future if we ever need created_by for a single rule.

mattrobenolt replied:
@wedamija i wonder if we can add a lint rule or something to complain and stop using .first().

wedamija commented on src/sentry/api/serializers/models/alert_rule.py:
```diff
@@ -56,6 +56,7 @@ def serialize(self, obj, attrs, user):
             "includeAllProjects": obj.include_all_projects,
             "dateModified": obj.date_modified,
             "dateCreated": obj.date_added,
+            "createdBy": obj.created_by,
```
We should handle this in `get_attrs` since it'll make a separate query for every `Incident` row.

iProgramStuff replied:
Okay, I can bulk fetch it for this serializer, but I am thinking I will leave the property in so it could be used as a sort field.

wedamija commented on src/sentry/api/serializers/models/rule.py:
```diff
@@ -43,6 +43,7 @@ def serialize(self, obj, attrs, user):
             "frequency": obj.data.get("frequency") or Rule.DEFAULT_FREQUENCY,
             "name": obj.label,
             "dateCreated": obj.date_added,
+            "createdBy": obj.created_by,
```
Same thing about `get_attrs` here.

iProgramStuff committed:
Bulk fetching created_by

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/trackalertrulecreator

iProgramStuff committed:
missing return value

iProgramStuff committed:
Removing unneeded user context

iProgramStuff committed:
Adding migrations

iProgramStuff committed:
Fixing created_by implementation

markstory committed:
chore(attachments) Cleanup dangling event attachment files (#20069)

The previous deletion code didn't cascade into EventAttachment files
because event attachments were deleted with a bulk SQL query. This has
left many unreferenced files in production that we would like to purge.

This migration looks at files older than the oldest event attachment and
removes files if they are no longer referenced by an attachment. This is
slower but safer.

markstory commented on migrations_lockfile.txt:
```diff
@@ -10,7 +10,7 @@ auth: 0008_alter_user_username_max_length
 jira_ac: 0001_initial
 nodestore: 0001_initial
-sentry: 0094_ruleactivity
+sentry: 0094_ruleactvity
```
You have a typo in your migration name.

iProgramStuff committed:
Merging with master, fixing migration

iProgramStuff committed:
Adding tests for alert rule tracking

iProgramStuff committed:
Adding tests for Rule activity

iProgramStuff committed:
Fixing bug

iProgramStuff committed:
Reverts

iProgramStuff committed:
Fixing serializer bug

iProgramStuff committed:
Fixing serializer bug

wedamija reviewed (commented):
Tests should also be updated to check this new field

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/ruleserializeradditions

iProgramStuff committed:
Merge remote-tracking branch 'origin' into feat/ruleserializeradditions

iProgramStuff committed:
Adding test for rule created_by
